rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,386 @@
+
+* `tsh ls`
+* `tsh ssh`
+* `tsh scp`
+
+The awarded certificates should not provide access to roles usually awarded to the user, as these roles may provide more privileges than required for the requested command. However, some role provided fields should be included, such as `role.options.forward_agent` and `role.logins`. Additionally, the user should be notified of what command was requested and what permissions will be awarded to the headless session once approved.
+
+Example:
+
+```
+Incoming headless request for command "tsh ssh server01", to connect to node <server01_uuid>. Upon approval, the request will be awarded the following permissions:
+
+allow:
+  options:
+    max_session_ttl: 8h0m0s
+    forward_agent: true
+    port_forwarding: true
+    require_session_mfa: true
+  logins: ["ubuntu-user"]
+  resource_ids: ["<server01_uuid>"]
+  rules:
+    resources: ["nodes"]
+    verbs: ["read", "list"]
+```
+
+**Important note**: The server cannot guarantee that the client will actually use the resulting certificates for the `tsh` command advertised by the client. This means that a modified client could claim to need certificates for `tsh ls`, but then actually use the certificates for `tsh ssh`. In order to prevent potential permission workarounds, we will need to carefully consider how each supported command's awarded certificates may interact with other Teleport clients. For this reason, we may add more commands as they are requested, but each command may have significant security implications to consider. Each PR adding one or more new commands should include a security description and get a security review from the Teleport security team. This includes the commands listed above, as the most limited necessary permissions necessary to perform each command is not yet clear.
+
+#### Caveat
+
+Due to the implementation complexity of limiting certificates permissions by requested command, it will not be included in the feature's preview release.
+
+Additionally, some details, like how the server will encode the permissions into the certificate, are currently missing. This will be addressed in a follow up PR before the official release of this feature, accompanied by an RFD update.
+
+The feature will still be limited to `tsh ls`, `tsh ssh`, and `tsh scp` so that we don't need to remove functionality later.
+
+#### Conclusion
+
+With the design principles above, the only possible attack would be for the attacker to issue the headless request themselves and trick the user into verifying the request with MFA, with a direct phishing attack. This phishing attack should also be mitigated by notifying the user of the command requested and the permissions that will be awarded upon approval.
```

Comment:
> This phishing attack should also be mitigated by notifying the user of the command requested and the permissions that will be awarded upon approval

I'm still concerned about phishing, I don't think it's a stretch that folks would approve requests without reading them, or get confused by a multitude of requests.

I'm not sure what we could do, so no action required here.

Reply:
That's fair. We could remove the "Headless auth requests" page from the proxy and only allow it to be completed from the generated link, keeping it more in line with standard SSO flow. The extra page is just a potential UX improvement for when a user is creating back to back headless authentication requests.